source/base/Promise.ooc

```diff
@@ -11,53 +12,59 @@ Promise: abstract class {
 	_state : _PromiseState
 	init: func
 	wait: abstract func -> Bool
+	wait: abstract func ~timeout (seconds: Double) -> Bool
 	cancel: virtual func -> Bool { false }
 	start: static func (action: Func) -> This {
 		_ThreadPromise new(action)
 	}
+	operator + (other: This) -> PromiseCollector {
+		collector := PromiseCollector new()
+		collector add(this)
+		collector add(other)
+		collector
+	}
 }
 
 _ThreadPromise: class extends Promise {
 	_action: Func
 	_thread: Thread
-	_mutex := Mutex new()
+	_threadAlive := true
 	init: func (task: Func) {
 		super()
 		this _state = _PromiseState Unfinished
 		this _action = func {
 			task()
-			this _mutex lock()
 			if (this _state != _PromiseState Cancelled)
 				this _state = _PromiseState Finished
-			this _mutex unlock()
 		}
 		this _thread = Thread new(this _action)
 		this _thread start()
 	}
 	free: override func {
-		if (this _state == _PromiseState Unfinished)
+		if (this _threadAlive)
 			this _thread wait()
```

Comment:
I have to join the thread before I can free its reference, and that means waiting for it. Do you have another suggestion?